commerce-cloud-code/app_yourshophere/cartridge/controllers/Cart.js

```diff
@@ -67,7 +83,171 @@ server.use('Delete', (req, res, next) => {
+    const BasketMgr = require('dw/order/BasketMgr');
+    const Transaction = require('dw/system/Transaction');
+    const Resource = require('dw/web/Resource');
+    const URLUtils = require('dw/web/URLUtils');
+    const ShippingMgr = require('dw/order/ShippingMgr');
+    const basket = BasketMgr.getCurrentBasket();
+    const shippingMethodID = req.httpParameterMap.shippingMethod.stringValue;
+    const redirectUrl = request.httpParameterMap.hx.submitted ? URLUtils.url('Cart-Show', 'hx', request.httpParameterMap.hx.stringValue || '') : URLUtils.url('Cart-Show');
+
+    if (!basket) {
+        // Handle the case where there is no basket
+        res.redirect(redirectUrl.append('error', Resource.msg('error.basket.missing', 'cart', null)));
+        return next();
+    }
+
+    const shipment = basket.getDefaultShipment();
+
+    if (!shipment) {
+        // Handle the case where there is no shipment
+        res.redirect(redirectUrl.append('error', Resource.msg('error.shipment.missing', 'cart', null)));
+        return next();
+    }
+
+    if (!shippingMethodID) {
+        // Handle the case where shippingMethod parameter is missing
+        res.redirect(redirectUrl.append('error', Resource.msg('error.shippingmethodid.missing', 'cart', null)));
+        return next();
+    }
+
+    const shippingMethod = ShippingMgr.getAllShippingMethods().toArray().filter((method) => method.ID === shippingMethodID).pop();
+
+    if (!shippingMethod) {
+        // Handle the case where the shipping method is not found
+        res.redirect(redirectUrl.append('error', Resource.msg('error.shippingmethod.notfound', 'cart', null)));
+        return next();
+    }
+
+    Transaction.wrap(() => {
+        shipment.setShippingMethod(shippingMethod);
+    });
+
+    // Redirect to the cart page with a success message
+    res.redirect(redirectUrl.append('success', Resource.msg('info.shipping.updated', 'cart', null)));
+    return next();
+});
+
+server.post('AddCoupon', server.middleware.https, (req, res, next) => {
+    const BasketMgr = require('dw/order/BasketMgr');
+    const URLUtils = require('dw/web/URLUtils');
+    const Transaction = require('dw/system/Transaction');
+
+    const CouponStatusCodes = require('dw/campaign/CouponStatusCodes');
+    const currentBasket = BasketMgr.getCurrentOrNewBasket();
+    const couponCode = req.httpParameterMap.couponCode.stringValue;
+    const redirectUrl = request.httpParameterMap.hx.submitted ? URLUtils.url('Cart-Show', 'hx', request.httpParameterMap.hx.stringValue || '') : URLUtils.url('Cart-Show');
+
+    if (!currentBasket) {
+        // Could log this error for monitoring
+    }
+
+    if (!couponCode) {
+        // Could log this error
+    }
+
+    try {
+        Transaction.wrap(() => {
+            currentBasket.createCouponLineItem(couponCode, true);
+        });
+    } catch (e) {
+        if (e.errorCode === CouponStatusCodes.COUPON_CODE_ALREADY_IN_BASKET) {
+            // ...
```

Comment:
please throw an error for now so it is obvious error handling need to be build. This silent treatment would potentially cause long lasting headscratching and debugging